### PR TITLE
Drop Q_FOREACH

### PR DESCRIPTION
**Pull request by luis-pereira**
Replaced by the ranged for.
Q_FOREACH will be deprecated soon. Enforced with the Q_NO_FOREACH
definition.

### CODE REVIEW

**Comment by tsujan** [comment]
I add that all cases *without* `static_cast` are correct.

**Comment by luis-pereira** [comment]
@tsujan Ignore the `static_cast`. It's `const_cast` obviously. Thanks for catching it.

**Comment by tsujan** [comment]
> Thanks for catching it.

Actually, I didn't ;) --  just checked the lines without it.

**Comment by tsujan** [comment]
And am I wrong, thinking that `static_cast` is correct?

**Comment by luis-pereira** [comment]
@tsujan Correct, yes.  The `const_cast` is more appropriated in this use case. We don't need a `static_cast` and a `const_cast` allows the compiler to catch more errors.

**Comment by luis-pereira** [comment]
@tsujan What I mean is: if a `const_cast` is enough (in this case it is) there's no need to use a `static_cast`.

**Comment by tsujan** [comment]
@luis-pereira I think `static_cast` is a little safer in general -- but, perhaps, not here,. However, if `const_cast` catches more errors, I'll have no objection to it either.

**Comment by luis-pereira** [comment]
@tsujan `const_cast` catches more errors, no doubt.
PR updated.

**Comment by palinek** [comment]
How about requiring Qt v5.7 and using the `qAsConst()`?

**Comment by tsujan** [comment]
> How about requiring Qt v5.7 and using the qAsConst()?

It came to my mind too -- when the container can't be declared as const -- but later thought, "they deprecated `Q_FOREACH`, so they might do the same thing with `qAsConst`.."

**Comment by luis-pereira** [comment]
If requiring Qt 5.7 is fine from a packagers/distro perspective, I'm fine with it.
On the technical side `qAsConst()` is syntactic sugar. It performs a `const_cast<const T&>`. `qAsConst()` is the Qt implementation of the new C++17 `std::as_const()`. When `std::as_const()` becomes available in Qt's targets compilers, it will most likely be deprecated. But that may take a while. When it becomes deprecated the transition will be simple because `qAsConst()` implementation follows `std::as_const()` on purpose.

**Comment by agaida** [comment]
> https://packages.debian.org/stretch/qt5-default
Even debian stable has 5.7.1 - so i would give it a go

**Comment by luis-pereira** [comment]
So... let's go with `qAsConst()` and requiring Qt 5.7. Comments wellcome.

**Comment by tsujan** [comment]
@luis-pereira As you mentioned, when it's going to be deprecated, it'll be easy to replace it. On the other hand, `std::as_const()` may not be available everywhere yet. So, I agree.

**Comment by palinek** [comment]
> So... let's go with qAsConst() and requiring Qt 5.7. Comments wellcome.

Full ack.

**Comment by agaida** [comment]
Full ack.

File: src/filesearchdialog.cpp [hunk 1]
```diff
@@ -144,7 +144,8 @@ void FileSearchDialog::onAddPath() {
 
 void FileSearchDialog::onRemovePath() {
     // remove selected items
-    Q_FOREACH(QListWidgetItem* item, ui->listView->selectedItems()) {
+    const auto itemList = ui->listView->selectedItems();
+    for(QListWidgetItem* const item : qAsConst(itemList)) {
```
Comment (tsujan) [inline src/filesearchdialog.cpp]: `qAsConst()` doesn't seem needed here.

File: src/dirtreemodelitem.cpp [hunk 2]
```diff
@@ -343,7 +343,7 @@ DirTreeModelItem* DirTreeModelItem::childFromName(const char* utf8_name, int* po
 DirTreeModelItem* DirTreeModelItem::childFromPath(Fm::FilePath path, bool recursive) const {
     Q_ASSERT(path != nullptr);
 
-    Q_FOREACH(DirTreeModelItem* item, children_) {
+    for(DirTreeModelItem* const item : const_cast<const std::vector<DirTreeModelItem*>&>(children_)) {
```
Comment (tsujan) [inline src/dirtreemodelitem.cpp]: Shouldn't this be like `hiddenChildren_` above?

**Comment by luis-pereira** [comment]
@tsujan Updated. Thanks.

**Comment by tsujan** [comment]
Very GTM.

**Pull request merged**